File: showcase/integrations/langgraph-typescript/src/agent/tool-rendering.ts
/**
 * Tool Rendering agent -- TypeScript port of tool_rendering_agent.py.
 *
 * Backs the tool-rendering demos:
 *   - tool-rendering-default-catchall  (no frontend renderers)
 *   - tool-rendering-custom-catchall   (wildcard renderer on frontend)
 *   - tool-rendering                   (per-tool + catch-all on frontend)
 *
 * All cells share this backend -- they differ only in how the frontend
 * renders the same tool calls.
 */

import { makeChatOpenAI } from "./openai-headers";

// @region[weather-tool-backend]
import { z } from "zod";
import type { RunnableConfig } from "@langchain/core/runnables";
import { tool } from "@langchain/core/tools";
import { ToolNode } from "@langchain/langgraph/prebuilt";
import type { AIMessage } from "@langchain/core/messages";
import { SystemMessage } from "@langchain/core/messages";
import {
  Annotation,
  MemorySaver,
  START,
  StateGraph,
} from "@langchain/langgraph";
import {
  convertActionsToDynamicStructuredTools,
  CopilotKitStateAnnotation,
} from "@copilotkit/sdk-js/langgraph";

// ---------------------------------------------------------------------------
// 1. Agent state -- extends CopilotKit state annotation
// ---------------------------------------------------------------------------

const AgentStateAnnotation = Annotation.Root({
  ...CopilotKitStateAnnotation.spec,
});

export type AgentState = typeof AgentStateAnnotation.State;

// ---------------------------------------------------------------------------
// 2. System prompt -- matches LGP exactly
// ---------------------------------------------------------------------------

const SYSTEM_PROMPT =
  "You are a travel & lifestyle concierge. Use the mock tools for " +
  "weather, flights, stock prices, or d20 rolls when the user asks; " +
  "otherwise reply in plain text. For flights, default origin to 'SFO' " +
  "if the user only names a destination. Call multiple tools in one " +
  "turn if asked. After tools return, summarize in one short sentence. " +
  "Never fabricate data a tool could provide.";

// ---------------------------------------------------------------------------
// 3. Tools -- aligned with LGP tool definitions
// ---------------------------------------------------------------------------

const getWeather = tool(
  async ({ location }) => ({
    city: location,
    temperature: 68,
    humidity: 55,
    wind_speed: 10,
    conditions: "Sunny",
  }),
  {
    name: "get_weather",
    description: "Get the current weather for a given location.",
    schema: z.object({
      location: z.string().describe("City name"),
    }),
  },
);
// @endregion[weather-tool-backend]

const searchFlights = tool(
  async ({ origin, destination }) => ({
    origin,
    destination,
    flights: [
      {
        airline: "United",
        flight: "UA231",
        depart: "08:15",
        arrive: "16:45",
        price_usd: 348,
      },
      {
        airline: "Delta",
        flight: "DL412",
        depart: "11:20",
        arrive: "19:55",
        price_usd: 312,
      },
      {
        airline: "JetBlue",
        flight: "B6722",
        depart: "17:05",
        arrive: "01:30",
        price_usd: 289,
      },
    ],
  }),
  {
    name: "search_flights",
    description:
      "Search mock flights from an origin airport to a destination airport.",
    schema: z.object({
      origin: z.string().describe("Origin airport code"),
      destination: z.string().describe("Destination airport code"),
    }),
  },
);

const getStockPrice = tool(
  async ({ ticker, price_usd, change_pct }) => {
    const randInt = (lo: number, hi: number) =>
      Math.floor(Math.random() * (hi - lo + 1)) + lo;
    const sign = Math.random() < 0.5 ? -1 : 1;
    return {
      ticker: ticker.toUpperCase(),
      price_usd:
        price_usd != null
          ? Math.round(price_usd * 100) / 100
          : Math.round((100 + randInt(0, 400) + randInt(0, 99) / 100) * 100) /
            100,
      change_pct:
        change_pct != null
          ? Math.round(change_pct * 100) / 100
          : Math.round(sign * (randInt(0, 300) / 100) * 100) / 100,
    };
  },
  {
    name: "get_stock_price",
    description:
      "Get a mock current price for a stock ticker.\n\n" +
      "The optional `price_usd` and `change_pct` arguments let the LLM (or " +
      "aimock fixture) script a deterministic ticker quote for testing -- " +
      "when supplied, the tool echoes them back verbatim. When omitted (or " +
      "null), the tool returns mock random values. Mirrors the " +
      "deterministic-`value` pattern on `roll_d20`.",
    schema: z.object({
      ticker: z.string().describe("Stock ticker symbol"),
      price_usd: z
        .number()
        .optional()
        .describe(
          "Deterministic price override for testing (echoed back verbatim)",
        ),
      change_pct: z
        .number()
        .optional()
        .describe(
          "Deterministic change-pct override for testing (echoed back verbatim)",
        ),
    }),
  },
);

const rollD20 = tool(
  async ({ value }) => {
    const rolled =
      typeof value === "number" && value >= 1 && value <= 20
        ? value
        : Math.floor(Math.random() * 20) + 1;
    return { sides: 20, value: rolled, result: rolled };
  },
  {
    name: "roll_d20",
    description: "Roll a 20-sided die.",
    schema: z.object({
      value: z
        .number()
        .int()
        .optional()
        .describe(
          "Deterministic override for the roll result (used by test fixtures)",
        ),
    }),
  },
);

const tools = [getWeather, searchFlights, getStockPrice, rollD20];

// ---------------------------------------------------------------------------
// 4. Chat node -- binds backend + frontend tools, invokes the model
// ---------------------------------------------------------------------------

async function chatNode(state: AgentState, config: RunnableConfig) {
  const model = makeChatOpenAI(config, { model: "gpt-5.4" });

  const modelWithTools = model.bindTools!([
    ...convertActionsToDynamicStructuredTools(state.copilotkit?.actions ?? []),
    ...tools,
  ]);

  const systemMessage = new SystemMessage({ content: SYSTEM_PROMPT });

  const response = await modelWithTools.invoke(
    [systemMessage, ...state.messages],
    config,
  );

  // Normalize tool calls that streamed in as `additional_kwargs.tool_calls`
  // but failed to promote to the parsed top-level `tool_calls`.
  //
  // When the model returns BOTH assistant content AND a tool call in one
  // message (as the tool-rendering fixtures do — "Looking up the weather…"
  // + get_weather), langchain-js's streamed-chunk reassembly can leave the
  // fully-formed tool call only under `additional_kwargs.tool_calls` while
  // the top-level `tool_calls` stays empty (the intermediate chunks surface
  // as `invalid_tool_calls` with "Malformed args" during accumulation), and
  // the merged result is a bare "generic" message rather than an AIMessage.
  // `shouldContinue`, `ToolNode`, and the AG-UI TOOL_CALL_* event emitter all
  // key off a proper AIMessage with top-level `tool_calls`, so the unparsed
  // result silently drops the tool call: the graph ends at chat_node (or
  // ToolNode rejects the non-AIMessage), the tool never runs, and the
  // frontend renders no card. LGP-python's `create_agent` parses the same
  // fixture response correctly, so this restores TS↔Python parity.
  return { messages: normalizeAssistantMessage(response) };
}

/**
 * Ensure the chat model's response is a well-formed `AIMessage` whose tool
 * calls live in the top-level `tool_calls` array. When langchain-js's
 * streamed reassembly leaves tool calls only under
 * `additional_kwargs.tool_calls` (and/or produces a non-AIMessage), rebuild
 * a clean `AIMessage` that preserves the content and promotes those tool
 * calls. A no-op passthrough when the response already parses correctly.
 */
function normalizeAssistantMessage(response: AIMessage): AIMessage {
  const alreadyHasToolCalls =
    Array.isArray(response.tool_calls) && response.tool_calls.length > 0;
  const isAIMessage = response instanceof AIMessage;
  if (alreadyHasToolCalls && isAIMessage) return response;

  const rawToolCalls = (
    response.additional_kwargs as { tool_calls?: unknown } | undefined
  )?.tool_calls;

  const parsed: NonNullable<AIMessage["tool_calls"]> = [];
  if (Array.isArray(rawToolCalls)) {
    for (const raw of rawToolCalls) {
      const fn = (raw as { function?: { name?: string; arguments?: string } })
        ?.function;
      const id = (raw as { id?: string })?.id;
      const name = fn?.name;
      if (!name) continue;
      let args: Record<string, unknown> = {};
      if (typeof fn?.arguments === "string" && fn.arguments.length > 0) {
        try {
          args = JSON.parse(fn.arguments);
        } catch {
          // Leave args empty when the accumulated JSON is unparseable
          // rather than dropping the whole tool call.
          args = {};
        }
      }
      parsed.push({ id, name, args, type: "tool_call" });
    }
  }

  // Nothing to reconcile and it's already an AIMessage — leave it untouched.
  if (parsed.length === 0 && isAIMessage) return response;

  return new AIMessage({
    id: response.id,
    content: response.content ?? "",
    tool_calls: parsed.length > 0 ? parsed : response.tool_calls,
    additional_kwargs: response.additional_kwargs,
    response_metadata: response.response_metadata,
  });
}

// ---------------------------------------------------------------------------
// 5. Routing -- send tool calls to tool_node unless they're CopilotKit
//    frontend actions.
// ---------------------------------------------------------------------------

function shouldContinue({ messages, copilotkit }: AgentState) {
  const lastMessage = messages[messages.length - 1] as AIMessage;

  if (lastMessage.tool_calls?.length) {
    const actions = copilotkit?.actions;
    const toolCallName = lastMessage.tool_calls![0].name;

    if (!actions || actions.every((action) => action.name !== toolCallName)) {
      return "tool_node";
    }
  }

  return "__end__";
}

// ---------------------------------------------------------------------------
// 6. Compile the graph
// ---------------------------------------------------------------------------

const workflow = new StateGraph(AgentStateAnnotation)
  .addNode("chat_node", chatNode)
  .addNode("tool_node", new ToolNode(tools))
  .addEdge(START, "chat_node")
  .addEdge("tool_node", "chat_node")
  .addConditionalEdges("chat_node", shouldContinue as any);

const memory = new MemorySaver();

export const graph = workflow.compile({
  checkpointer: memory,
});

File: showcase/integrations/crewai-conversational-flows/src/agents/tool_rendering.py
"""CrewAI Flow backing the Tool Rendering demo.

The default ``ChatWithCrewFlow`` (used for the catch-all
``LatestAiDevelopment`` crew on "/") executes backend tools internally
without emitting AG-UI ``TOOL_CALL_START`` / ``TOOL_CALL_END`` events.
The frontend's ``useRenderTool`` hook never sees the tool call and
therefore never renders the WeatherCard.

This module bypasses the crew flow and uses a raw CrewAI ``Flow`` with
``copilotkit_stream`` -- which DOES emit AG-UI tool-call events for
every tool call in the LLM response -- so the frontend receives the
tool call and the registered ``useRenderTool`` renderer fires.

The flow mirrors the LangGraph-Python reference: it makes LLM calls in
a loop, executing backend tools (``get_weather``) locally and streaming
every response (text or tool call) through ``copilotkit_stream``.
"""

from __future__ import annotations

import json
import uuid
from typing import List, Optional

from crewai.flow.flow import Flow, start
from litellm import acompletion
from pydantic import Field

from ag_ui_crewai import (
    CopilotKitState,
    copilotkit_emit_tool_result,
    copilotkit_stream,
)

from tools import get_weather_impl, roll_dice_impl


# ---------------------------------------------------------------------------
# State
# ---------------------------------------------------------------------------


class ToolRenderingState(CopilotKitState):
    """Minimal state -- just the conversation messages."""

    pass


# ---------------------------------------------------------------------------
# Tool schema (LiteLLM/OpenAI tool format)
# ---------------------------------------------------------------------------

GET_WEATHER_TOOL = {
    "type": "function",
    "function": {
        "name": "get_weather",
        "description": (
            "Get current weather for a location.  Always call this tool "
            "when the user asks about weather.  Ensure the location is "
            "fully spelled out (e.g. 'San Francisco', not 'SF')."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "location": {
                    "type": "string",
                    "description": "The city or location to get weather for.",
                }
            },
            "required": ["location"],
        },
    },
}


GET_STOCK_PRICE_TOOL = {
    "type": "function",
    "function": {
        "name": "get_stock_price",
        "description": (
            "Get a mock current price for a stock ticker.  Always call "
            "this tool when the user asks about a stock price or quote. "
            "Pass the ticker symbol verbatim (e.g. 'AAPL')."
        ),
        "parameters": {
            "type": "object",
            "properties": {
                "ticker": {
                    "type": "string",
                    "description": "The stock ticker symbol (e.g. 'AAPL').",
                },
                "price_usd": {
                    "type": "number",
                    "description": (
                        "Optional deterministic price to echo back. When "
                        "omitted, a mock price is returned."
                    ),
                },
                "change_pct": {
                    "type": "number",
                    "description": (
                        "Optional deterministic percent-change to echo back. "
                        "When omitted, a mock change is returned."
                    ),
                },
            },
            "required": ["ticker"],
        },
    },
}

SEARCH_FLIGHTS_TOOL = {
    "type": "function",
    "function": {
        "name": "search_flights",
        "description": "Search mock flights between two airports.",
        "parameters": {
            "type": "object",
            "properties": {
                "origin": {"type": "string"},
                "destination": {"type": "string"},
            },
            "required": ["origin", "destination"],
        },
    },
}

ROLL_D20_TOOL = {
    "type": "function",
    "function": {
        "name": "roll_d20",
        "description": "Roll a 20-sided die, optionally using a test value.",
        "parameters": {
            "type": "object",
            "properties": {"value": {"type": "integer", "minimum": 1, "maximum": 20}},
        },
    },
}


def get_stock_price_impl(
    ticker: str,
    price_usd: float | None = None,
    change_pct: float | None = None,
) -> dict:
    """Return mock stock quote for the given ticker.

    Mirrors the LangGraph-Python `get_stock_price` tool shape so the
    aimock fixtures and `d5-tool-rendering-custom-catchall` probe see
    identical tool-result payloads across integrations. When
    `price_usd`/`change_pct` are supplied (e.g. by the aimock fixture
    "Quote AAPL through the wildcard renderer"), they're echoed back
    verbatim for deterministic assertions.
    """
    import random as _random

    rng = _random.Random(ticker.lower())
    return {
        "ticker": ticker.upper(),
        "price_usd": (
            round(float(price_usd), 2)
            if price_usd is not None
            else round(100 + rng.randint(0, 400) + rng.randint(0, 99) / 100, 2)
        ),
        "change_pct": (
            round(float(change_pct), 2)
            if change_pct is not None
            else round(rng.choice([-1, 1]) * (rng.randint(0, 300) / 100), 2)
        ),
    }


# ---------------------------------------------------------------------------
# Flow
# ---------------------------------------------------------------------------

_SYSTEM_PROMPT = (
    "You are a helpful, concise assistant.  When the user asks about "
    "weather for a location, call the `get_weather` tool.  When the "
    "user asks about a stock price or quote, call the `get_stock_price` "
    "tool with the ticker symbol.  After receiving any tool result, "
    "summarise it in a short sentence."
)

# Maximum LLM round-trips per user turn (prevents infinite loops).
_MAX_ITERATIONS = 8


def search_flights_impl(origin: str, destination: str) -> dict:
    return {
        "origin": origin,
        "destination": destination,
        "flights": [
            {
                "airline": "United",
                "flight": "UA231",
                "depart": "08:15",
                "arrive": "16:45",
                "price_usd": 348,
            },
            {
                "airline": "Delta",
                "flight": "DL412",
                "depart": "11:20",
                "arrive": "19:55",
                "price_usd": 312,
            },
            {
                "airline": "JetBlue",
                "flight": "B6722",
                "depart": "17:05",
                "arrive": "01:30",
                "price_usd": 289,
            },
        ],
    }


class ToolRenderingFlow(Flow[ToolRenderingState]):
    """Chat flow that streams tool calls to the frontend for rendering."""

    @start()
    async def chat(self) -> None:
        system_message = {
            "role": "system",
            "content": _SYSTEM_PROMPT,
            "id": str(uuid.uuid4()) + "-system",
        }

        # Frontend-registered actions + our backend get_weather / get_stock_price tools.
        tools = [
            *self.state.copilotkit.actions,
            GET_WEATHER_TOOL,
            GET_STOCK_PRICE_TOOL,
            SEARCH_FLIGHTS_TOOL,
            ROLL_D20_TOOL,
        ]

        for _iteration in range(_MAX_ITERATIONS):
            messages = [system_message, *self.state.messages]

            response = await copilotkit_stream(
                await acompletion(
                    model="openai/gpt-5.4",
                    messages=messages,
                    tools=tools,
                    parallel_tool_calls=False,
                    stream=True,
                )
            )

            message = response.choices[0].message
            self.state.messages.append(message)

            tool_calls = message.get("tool_calls") or []
            if not tool_calls:
                # No tool calls -- the LLM produced a text response.
                return

            for tool_call in tool_calls:
                tool_call_id = tool_call["id"]
                tool_name = tool_call["function"]["name"]

                if tool_name == "get_weather":
                    try:
                        args = json.loads(tool_call["function"]["arguments"] or "{}")
                    except json.JSONDecodeError:
                        args = {}
                    location = args.get("location", "Unknown")
                    result = get_weather_impl(location)
                    result_str = json.dumps(result)

                    self.state.messages.append(
                        {
                            "role": "tool",
                            "content": result_str,
                            "tool_call_id": tool_call_id,
                        }
                    )
                    await copilotkit_emit_tool_result(tool_call_id, result_str)
                elif tool_name == "get_stock_price":
                    try:
                        args = json.loads(tool_call["function"]["arguments"] or "{}")
                    except json.JSONDecodeError:
                        args = {}
                    ticker = args.get("ticker", "UNKNOWN")
                    price_usd = args.get("price_usd")
                    change_pct = args.get("change_pct")
                    stock_result = get_stock_price_impl(
                        ticker, price_usd=price_usd, change_pct=change_pct
                    )
                    self.state.messages.append(
                        {
                            "role": "tool",
                            "content": json.dumps(stock_result),
                            "tool_call_id": tool_call_id,
                        }
                    )
                    await copilotkit_emit_tool_result(
                        tool_call_id, json.dumps(stock_result)
                    )
                elif tool_name == "search_flights":
                    try:
                        args = json.loads(tool_call["function"]["arguments"] or "{}")
                    except json.JSONDecodeError:
                        args = {}
                    flight_result = search_flights_impl(
                        args.get("origin", "SFO"), args.get("destination", "JFK")
                    )
                    result_str = json.dumps(flight_result)
                    self.state.messages.append(
                        {
                            "role": "tool",
                            "content": result_str,
                            "tool_call_id": tool_call_id,
                        }
                    )
                    await copilotkit_emit_tool_result(tool_call_id, result_str)
                elif tool_name == "roll_d20":
                    try:
                        args = json.loads(tool_call["function"]["arguments"] or "{}")
                    except json.JSONDecodeError:
                        args = {}
                    scripted = args.get("value")
                    dice_result = (
                        {"sides": 20, "value": scripted, "result": scripted}
                        if isinstance(scripted, int) and 1 <= scripted <= 20
                        else {**roll_dice_impl(20), "value": None}
                    )
                    if dice_result["value"] is None:
                        dice_result["value"] = dice_result["result"]
                    result_str = json.dumps(dice_result)
                    self.state.messages.append(
                        {
                            "role": "tool",
                            "content": result_str,
                            "tool_call_id": tool_call_id,
                        }
                    )
                    await copilotkit_emit_tool_result(tool_call_id, result_str)
                else:
                    # The browser owns frontend-tool execution. End this run
                    # after streaming the call; CopilotKit resumes with the
                    # real tool result in the next request.
                    return

            # Loop back to call the LLM again with the tool results.


# Module-level singleton -- deepcopied per request by the endpoint.
tool_rendering_flow = ToolRenderingFlow()
